libfwupdplugin/fu-common.c

```diff
@@ -1941,6 +1941,95 @@ fu_common_strnsplit(const gchar *str, gsize sz, const gchar *delimiter, gint max
 	return g_strsplit(str, delimiter, max_tokens);
 }
 
+/**
+ * fu_common_strnsplit_full:
+ * @str: a string to split
+ * @sz: size of @str, or -1 for unknown
+ * @delimiter: a string which specifies the places at which to split the string
+ * @max_tokens: the maximum number of pieces to split @str into, or 0 for no limit
+ * @callback: (scope call): a #FuCommonStrsplitCbFunc.
+ * @user_data: user data
+ * @error: (nullable): optional return location for an error
+ *
+ * Splits the string, calling the given function for each
+ * of the tokens found. If any @callback returns %FALSE scanning is aborted.
+ *
+ * Use this function in preference to fu_common_strnsplit() when the input file is untrusted,
+ * and you don't want to allocate a GStrv with billions of one byte items.
+ *
+ * Returns: %TRUE if no @callback returned FALSE
+ *
+ * Since: 1.7.0
+ */
+gboolean
+fu_common_strnsplit_full(const gchar *str,
+			 gssize sz,
+			 const gchar *delimiter,
+			 guint max_tokens,
+			 FuCommonStrsplitCbFunc callback,
+			 gpointer user_data,
+			 GError **error)
+{
+	gsize delimiter_sz;
+	gsize str_sz;
+	guint found_idx = 0;
+	guint token_idx = 0;
+
+	g_return_val_if_fail(str != NULL, FALSE);
+	g_return_val_if_fail(delimiter != NULL, FALSE);
+	g_return_val_if_fail(callback != NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	/* make known */
+	str_sz = sz != -1 ? (gsize)sz : strlen(str);
+	delimiter_sz = strlen(delimiter);
+	if (delimiter_sz == 0)
+		return TRUE;
+
+	/* can never match */
+	if (delimiter_sz > str_sz)
+		return TRUE;
```

Comment:
shouldn't these be failures - IE failure to split that the caller should know about?

Reply:
I think `delimiter_sz == 0` should probably be a return_val_if_fail programmer critical. I'm not sure about "delimiter_sz > str_sz" -- you can argue it both ways.

Reply:
Fixed both.